utils/ensemble_tools.py

```diff
@@ -0,0 +1,166 @@
+    combinations = list(itertools.combinations(list_all, len(parameter_dict.keys())))
+    all_configs = []
+    for comb in combinations:
+        record_comb = True
+        for i in range(len(comb) - 1):
+            search_term = comb[i][: len(comb[i]) - comb[i][::-1].find("_")]
+            for j in range(i + 1, len(comb)):
+                if comb[j].startswith(search_term):
+                    record_comb = False
+                    break
+        if record_comb:
+            all_configs.append(comb)
+
+    # If destructive, let's delete the whole tree if it already exists
+    if destructive and os.path.isdir(sweep_path):
+        shutil.rmtree(sweep_path)
+    # Check whether the copy directory exists
+    copy_files = os.path.isdir(copy_dir)
+    # Now loop over each combination, create a directory, copy the
+    # configs and conditions, and edit the configuation to contain
+    # the correct parameters.
+    for comb in all_configs:
+        dir_name = "-".join(comb)
+        dir_path = Path(sweep_path) / "SWEEP" / dir_name
+        make_directory(dir_path, destructive)
+        if copy_files:
+            copy_dir_contents(dir_path, copy_dir)
+        # Create parameter map for conditions encoding
+        parameters = {}
+        for parameter in parameter_dict.keys():
+            param_index = [item for item in comb if item.startswith(parameter)][
+                0
+            ].split("_")[-1]
+            parameters[parameter] = scan_points[parameter][int(param_index)]
+        if edit_file:
+            encode_conditions_file(Path(dir_path) / edit_file, parameters)
+
+
+def copy_dir_contents(dir_path, copy_dir):
+    """Copy the contents of copy_dir -> dir_path"""
+
+    shutil.copytree(copy_dir, dir_path, dirs_exist_ok=True)
+
+
+def make_directory(directory_name, destructive):
+    """Make the directory tree at directory_name."""
+
+    if os.path.isdir(directory_name):
+        if destructive:
+            shutil.rmtree(directory_name)
+        else:
+            msg = "Path already exists and not in destructive mode"
+            raise FileExistsError(msg)
+    os.makedirs(directory_name)
+
+
+def encode_conditions_file(in_file_name, param_dict):
+    """Encode a configuration file with a dict of input values"""
+
+    for param_name in param_dict.keys():
+        edit_parameter(in_file_name, param_name, param_dict[param_name])
+
+
+def edit_parameter(in_file_name, param, val):
+    """Edit a single parameter in the configuration
+    file to the desired value"""
+
+    print(f"Edit {param} : {val}")
+
+    data = ET.parse(in_file_name)
+    root = data.getroot()
+    parameters = root.find("CONDITIONS").find("PARAMETERS")
+    for element in parameters.iter("P"):
+        match = re.match(
+            r"\s*(?P<key>\w+)\s*=\s*(?P<value>[0-9.-]+)\s*",
+            element.text,
+        )
+        if match is None:
+            # This happens for the constructed elements, so skip them
+            continue
```

Comment:
Adding a `breakpoint()` here and running `make_sweep_dir.py` script locally suggests this if-block isn't entered in normal operation. What does 'constructed elements' in the comment here refer to? If we don't get a match I would be inclined to raise an error as that suggests we have a parameter defnition in a format we don't know how to handle.

Reply:
Ah I think I understand now - by 'constructed elements' you mean parameter definitions which involve other parameter values rather such as in the `hw.xml` file? Here we can probably just use a looser regex that will match these cases as well, particularly as we don't actually need to capture the actual original values here in a group (I did this in `nesopy` so we could record a dictionary of all parameter values including those not overriden). We only need to capture the key in a group so something like `r"\s*(?P<key>\w+)\s*="`  should work I think.

Reply:
It was indeed from those parameters in the ```hw.xml```. Changing the regex does indeed fix the problem, so I've changed that now.